### ISSUE TITLE
Empty array behavior with CollectionToArrayHeuristics

### ISSUE THREAD

**Issue by nicbec** [typetools/checker-framework#1929]
https://github.com/typetools/checker-framework/blob/6deea6321310e7b2d8eed197ef71e02f64230d81/checker/src/org/checkerframework/checker/nullness/CollectionToArrayHeuristics.java#L132

[CollectionToArrayHeuristics](https://checkerframework.org/api/org/checkerframework/checker/nullness/CollectionToArrayHeuristics.html) describes several heuristics for evaluating Collection.toArray. [Collection.toArray()](https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#toArray-T:A-) will possibly set the array element after the last value from the collection to null. This shouldn't actually be a problem in this case, because the EMPTY_MYCLASS_ARRAY is not large enough to be used directly, so Collection.toArray() will allocate a new array of just the right size.

```
private static final MyClass[] EMPTY_MYCLASS_ARRAY = new MyClass[0];
...
return result.values().toArray(EMPTY_MYCLASS_ARRAY);
```

Which does not satisfy 2. array creation tree of size 0, e.g. c.toArray(new String[0]), perhaps because the empty array is static/final? The following new allocation does:
```
return result.values().toArray(new MyClass[result.values().size()]);
```

We're not sure if this discrepancy is because CollectionToArrayHeuristics is operating directly on the abstract syntax tree level, rather than on the type level, or some other reason entirely. Perhaps the heuristic could be integrated with the [constant value checker](https://checkerframework.org/manual/#constant-value-checker) such as the ArrayLen annotation. For example if the checker framework inferred EMPTY_MYCLASS_ARRAY as ArrayLen(0) and CollectionToArrayHeuristics could check for ArrayLen(0). We aren't sure if this is a valid approach, or if our understanding of the behavior is correct. It would be great as a feature request if this is the case. Thanks!

**Comment by mernst** [comment]
Thanks for the bug report.
Here is a complete test case:

```java
import java.util.Collection;
import org.checkerframework.common.value.qual.ArrayLen;

public class CollectionToArrayTest {

    String[] works1(Collection<String> c) {
        return c.toArray(new String[0]);
    }

    private static final String @ArrayLen(0) [] EMPTY_STRING_ARRAY_2 = new String[0];

    String[] fails2(Collection<String> c) {
        return c.toArray(EMPTY_STRING_ARRAY_2);
    }

    private static final String[] EMPTY_STRING_ARRAY_3 = new String[0];

    String[] fails3(Collection<String> c) {
        return c.toArray(EMPTY_STRING_ARRAY_3);
    }
}
```

All three of the methods return an array that contains no null elements, but running
```javacheck -g CollectionToArrayTest.java -processor nullness```
produces type-checking errors
```
  found   : @Initialized @Nullable String @Initialized @NonNull []
  required: @Initialized @NonNull String @Initialized @NonNull []
```
for `fails2` and `fails3`.

The feature request is twofold:
 * The Nullness Checker should recognize Constant Value Checker annotations -- namely, `@ArrayLen(0)` -- rather than hard-coding tests for specific syntax.
 * The Nullness Checker should always run the Constant Value Checker, to validate its annotations.  One concern here is that this could slow down the Constant Value Checker, which is already fairly slow.  I would be curious to see timing results that indicate the slowdown -- maybe it's not a problem, or maybe it is.


(Minor request:  If possible, it the future could you please start out with concrete code and say what the type-checking result is?  Your message didn't say what the results were, nor give a compilable program that would have let me determine that for myself easily.  We want to help you, but having to figure this out delays our consideration of bug reports.  Thanks!)

**Comment by nicbec** [comment]
Thanks for the follow up, and for demonstrating the desired format for future requests. Noted!
